File: apps/windows/TeslaSync.App/feature-views/AIProviderSection/AIProviderSection.Model.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using TeslaSync.App.Core.Data.State;
using TeslaSync.App.Core.Notifications;

namespace TeslaSync.App.FeatureViews;

/// <summary>
/// The structured reason a provider validation was rejected — the native analogue of the web
/// <c>ValidateAiProviderReason</c> union (web/src/api/hooks/useAiSettings.ts). The wire codes mirror the
/// constants in <c>internal/api/ai_settings_validate_handler.go</c>; <see cref="Unknown"/> is the defensive
/// fallback the server may widen into. Pure data, unit-tested without a UI host.
/// </summary>
public enum AiProviderValidationReason
{
    /// <summary>Base URL resolved to a public address (local mode requires a private target).</summary>
    NotLocal,

    /// <summary>Malformed URL, DNS failure, or another generic local-validator rejection.</summary>
    Invalid,

    /// <summary>Request body carried <c>mode=off</c> or an unknown mode.</summary>
    BadMode,

    /// <summary>Request body was malformed.</summary>
    BadRequest,

    /// <summary>Cloud probe hit a provider name with no registered adapter.</summary>
    UnknownProvider,

    /// <summary>Cloud probe needs an API key (request omitted it and no saved key fallback).</summary>
    MissingApiKey,

    /// <summary>Azure flavor needs a resource endpoint.</summary>
    MissingBaseUrl,

    /// <summary>Azure OpenAI Service flavor needs a deployment name (or model) to route to.</summary>
    MissingDeployment,

    /// <summary>Provider returned 401/403 (bad key).</summary>
    Unauthorized,

    /// <summary>Provider returned 404 (bad URL or deployment slug).</summary>
    NotFound,

    /// <summary>Provider returned 5xx/429 or a transport failure (provider-side problem).</summary>
    UpstreamError,

    /// <summary>Probe exceeded the budget.</summary>
    Timeout,

    /// <summary>Fallback when the server omitted (or widened) the code.</summary>
    Unknown,
}

/// <summary>
/// Maps the backend's structured <c>code</c> string onto a typed <see cref="AiProviderValidationReason"/> —
/// the native mirror of the web <c>reasonFromCode</c> switch. Unknown / future codes narrow to
/// <see cref="AiProviderValidationReason.Unknown"/> so exhaustive handling keeps working.
/// </summary>
public static class AiProviderValidationReasons
{
    /// <summary>Resolves a wire <c>code</c> to its typed reason (null / unrecognised → <c>Unknown</c>).</summary>
    public static AiProviderValidationReason FromCode(string? code) => code switch
    {
        "not_local" => AiProviderValidationReason.NotLocal,
        "invalid" => AiProviderValidationReason.Invalid,
        "bad_mode" => AiProviderValidationReason.BadMode,
        "bad_request" => AiProviderValidationReason.BadRequest,
        "unknown_provider" => AiProviderValidationReason.UnknownProvider,
        "missing_api_key" => AiProviderValidationReason.MissingApiKey,
        "missing_base_url" => AiProviderValidationReason.MissingBaseUrl,
        "missing_deployment" => AiProviderValidationReason.MissingDeployment,
        "unauthorized" => AiProviderValidationReason.Unauthorized,
        "not_found" => AiProviderValidationReason.NotFound,
        "upstream_error" => AiProviderValidationReason.UpstreamError,
        "timeout" => AiProviderValidationReason.Timeout,
        _ => AiProviderValidationReason.Unknown,
    };
}

/// <summary>The lifecycle of a single validation attempt.</summary>
public enum AiProviderValidationStatus
{
    /// <summary>The probe succeeded (HTTP 200, <c>ok: true</c>).</summary>
    Ok,

    /// <summary>The validator rejected the configuration (HTTP 422 with a structured code + server message).</summary>
    Rejected,

    /// <summary>The request never produced a validation verdict (transport / 5xx fault).</summary>
    Faulted,
}

/// <summary>
/// The discriminated result of a provider validation — the native analogue of the web
/// <c>ValidateAiProviderResult</c>. A 200 yields <see cref="Status"/> <c>Ok</c> (with the optional
/// <see cref="PinnedIp"/> / <see cref="ProbedModel"/> the UI echoes); a 422 yields <c>Rejected</c> carrying the
/// server's structured <see cref="Reason"/> and verbatim <see cref="Message"/>; any other fault yields
/// <c>Faulted</c> carrying a classified <see cref="RepositoryError"/> the view localises. Pure data.
/// </summary>
public sealed record AiProviderValidationOutcome(
    AiProviderValidationStatus Status,
    string? PinnedIp = null,
    string? ProbedModel = null,
    string? Note = null,
    AiProviderValidationReason Reason = AiProviderValidationReason.Unknown,
    string Message = "",
    RepositoryError? Error = null)
{
    /// <summary>True when the provider was reachable.</summary>
    public bool IsOk => Status == AiProviderValidationStatus.Ok;

    /// <summary>A successful probe, echoing the pinned IP / probed model when the server supplied them.</summary>
    public static AiProviderValidationOutcome Success(string? pinnedIp, string? probedModel, string? note = null) =>
        new(AiProviderValidationStatus.Ok, NullIfBlank(pinnedIp), NullIfBlank(probedModel), NullIfBlank(note));

    /// <summary>A structured validator rejection (HTTP 422); the reason code drives the localised banner.</summary>
    public static AiProviderValidationOutcome Rejected(AiProviderValidationReason reason, string message = "") =>
        new(AiProviderValidationStatus.Rejected, Reason: reason, Message: message ?? string.Empty);

    /// <summary>A transport / server fault with no validation verdict; the view renders a localised message.</summary>
    public static AiProviderValidationOutcome Faulted(RepositoryError error)
    {
        ArgumentNullException.ThrowIfNull(error);
        return new AiProviderValidationOutcome(
            AiProviderValidationStatus.Faulted,
            Reason: AiProviderValidationReason.UpstreamError,
            Message: error.Message,
            Error: error);
    }

    /// <summary>
    /// Projects the HTTP-200 body into a success outcome. The validator always returns <c>ok: true</c> on a
    /// 200; a defensively-false body degrades to an <c>Unknown</c> rejection so the UI never shows a green
    /// banner for a not-ok payload.
    /// </summary>
    public static AiProviderValidationOutcome FromResponse(JsonElement body)
    {
        if (body.ValueKind != JsonValueKind.Object || !ReadBool(body, "ok"))
        {
            return Rejected(AiProviderValidationReason.Unknown);
        }

        return Success(ReadString(body, "pinned_ip"), ReadString(body, "probed_model"), ReadString(body, "note"));
    }

    private static string? NullIfBlank(string? value) => string.IsNullOrWhiteSpace(value) ? null : value;

    private static string? ReadString(JsonElement obj, string name) =>
        obj.ValueKind == JsonValueKind.Object
        && obj.TryGetProperty(name, out var prop)
        && prop.ValueKind == JsonValueKind.String
            ? prop.GetString()
            : null;

    private static bool ReadBool(JsonElement obj, string name) =>
        obj.ValueKind == JsonValueKind.Object
        && obj.TryGetProperty(name, out var prop)
        && prop.ValueKind == JsonValueKind.True;
}

/// <summary>
/// Builds the snake_case JSON body posted to <c>POST /settings/ai/validate-config</c> — the native mirror of
/// the web <c>runValidate</c> request shape. Local mode forwards only <c>mode</c> + <c>provider</c> +
/// <c>base_url</c> (the validator is provider-agnostic there); cloud mode forwards the full configuration and
/// omits <c>api_key</c> when the user left it blank so the backend falls back to the saved encrypted value
/// rather than clobbering it. Pure data so the wire contract is unit-tested without a UI host.
/// </summary>
public static class AiProviderValidationPayload
{
    /// <summary>Constructs the request body for the supplied draft + transport mode.</summary>
    public static JsonObject Build(AiProviderDraft draft, bool isCloud)
    {
        ArgumentNullException.ThrowIfNull(draft);

        if (!isCloud)
        {
            return new JsonObject
            {
                ["mode"] = "local",
                ["provider"] = draft.Provider,
                ["base_url"] = draft.BaseUrl,
            };
        }

        var body = new JsonObject
        {
            ["mode"] = "cloud",
            ["provider"] = draft.Provider,
            ["base_url"] = draft.BaseUrl,
            ["model"] = draft.Model,
            ["api_version"] = draft.ApiVersion,
            ["flavor"] = draft.Flavor,
            ["deployment"] = draft.Deployment,
            ["embedding_model"] = draft.EmbeddingModel,
            ["embedding_deployment"] = draft.EmbeddingDeployment,
        };

        // Only forward api_key when the user actually typed one; an empty field lets the backend reuse the
        // saved (encrypted) value rather than overwriting it with "".
        if (!string.IsNullOrWhiteSpace(draft.ApiKey))
        {
            body["api_key"] = draft.ApiKey;
        }

        return body;
    }
}

/// <summary>
/// Builds the localised banner copy for a finished validation — the native mirror of the web success-message
/// branch in <c>runValidate</c> plus the failure fallbacks. A pinned IP wins over a probed model wins over the
/// generic "reachable" line; a structured rejection shows the server's verbatim message (dynamic data, not a
/// UI literal); a transport fault shows a localised generic line. Interpolation follows the i18next
/// <c>{{token}}</c> convention so a translated resource and the English fallback share one shape. Pure logic.
/// </summary>
public static class AiProviderValidationCopy
{
    /// <summary>The success line for an OK outcome (pinned IP &gt; probed model &gt; generic).</summary>
    public static string Success(ILocalizer localizer, string? pinnedIp, string? probedModel)
    {
        ArgumentNullException.ThrowIfNull(localizer);

        if (!string.IsNullOrWhiteSpace(pinnedIp))
        {
            return Interpolate(
                localizer.GetString("ai.settings.validate.successPinned", "OK \u2014 pinned to {{ip}}"),
                "ip",
                pinnedIp!);
        }

        if (!string.IsNullOrWhiteSpace(probedModel))
        {
            return Interpolate(
                localizer.GetString("ai.settings.validate.successProbed", "OK \u2014 {{model}} reachable"),
                "model",
                probedModel!);
        }

        return localizer.GetString("ai.settings.validate.success", "OK \u2014 provider reachable");
    }

    /// <summary>
    /// The banner message for a failed outcome. A structured 422 rejection localises a clear line from its
    /// reason code (the native generated API client surfaces the structured <c>code</c>, not the server's
    /// free-form prose, so this is the i18n-correct equivalent of the web's verbatim server text); a
    /// transport / server fault localises a generic line.
    /// </summary>
    public static string Failure(ILocalizer localizer, AiProviderValidationOutcome outcome)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        ArgumentNullException.ThrowIfNull(outcome);

        if (outcome.Status == AiProviderValidationStatus.Rejected)
        {
            return ReasonMessage(localizer, outcome.Reason);
        }

        return localizer.GetString(
            "ai.settings.validate.faulted",
            "Validation could not complete. Check the API and try again.");
    }

    private static string ReasonMessage(ILocalizer localizer, AiProviderValidationReason reason) => reason switch
    {
        AiProviderValidationReason.NotLocal => localizer.GetString(
            "ai.settings.validate.reason.notLocal",
            "That address is not private. Local mode needs a loopback or private-network URL."),
        AiProviderValidationReason.Invalid => localizer.GetString(
            "ai.settings.validate.reason.invalid",
            "That URL could not be validated. Check it and try again."),
        AiProviderValidationReason.BadMode => localizer.GetString(
            "ai.settings.validate.reason.badMode",
            "Validation is unavailable for the current mode."),
        AiProviderValidationReason.BadRequest => localizer.GetString(
            "ai.settings.validate.reason.badRequest",
            "The validation request was malformed."),
        AiProviderValidationReason.UnknownProvider => localizer.GetString(
            "ai.settings.validate.reason.unknownProvider",
            "That provider has no adapter configured."),
        AiProviderValidationReason.MissingApiKey => localizer.GetString(
            "ai.settings.validate.reason.missingApiKey",
            "An API key is required to reach this provider."),
        AiProviderValidationReason.MissingBaseUrl => localizer.GetString(
            "ai.settings.validate.reason.missingBaseUrl",
            "A resource endpoint URL is required."),
        AiProviderValidationReason.MissingDeployment => localizer.GetString(
            "ai.settings.validate.reason.missingDeployment",
            "A deployment name (or model) is required to route the request."),
        AiProviderValidationReason.Unauthorized => localizer.GetString(
            "ai.settings.validate.reason.unauthorized",
            "The provider rejected the credentials."),
        AiProviderValidationReason.NotFound => localizer.GetString(
            "ai.settings.validate.reason.notFound",
            "The provider endpoint or deployment was not found."),
        AiProviderValidationReason.UpstreamError => localizer.GetString(
            "ai.settings.validate.reason.upstreamError",
            "The provider returned an error. Try again shortly."),
        AiProviderValidationReason.Timeout => localizer.GetString(
            "ai.settings.validate.reason.timeout",
            "The validation probe timed out."),
        _ => localizer.GetString("ai.settings.validate.reason.unknown", "Validation failed."),
    };

    private static string Interpolate(string template, string token, string value) =>
        template.Replace("{{" + token + "}}", value, StringComparison.Ordinal);
}

/// <summary>
/// Stable identity + localised title for the provider-configuration surface (P1/S11 registry). The
/// <see cref="Id"/> matches the web <c>data-testid</c> (<c>ai-provider-section</c>); the <see cref="Slug"/> is
/// the diagnostics surface name. Pure metadata.
/// </summary>
public static class AiProviderSectionRegistration
{
    /// <summary>The canonical surface id (matches the web test id).</summary>
    public const string Id = "ai-provider-section";

    /// <summary>The diagnostics surface slug.</summary>
    public const string Slug = "AIProviderSection";

    /// <summary>The localised section heading (web <c>ai.settings.provider.label</c>).</summary>
    public static string Title(ILocalizer localizer)
    {
        ArgumentNullException.ThrowIfNull(localizer);
        return localizer.GetString("ai.settings.provider.label", "Provider configuration");
    }
}

/// <summary>
/// PII-safe diagnostics for the provider surface (P1/S11 diagnostics contract). Records only the operational
/// <c>view.opened</c> event with the surface slug — never a provider name, key, URL or model — so a
/// diagnostics line can never leak operator-specific configuration. Thread-safe.
/// </summary>
public sealed class AiProviderSectionDiagnostics
{
    private readonly Action<string>? _sink;
    private long _viewsOpened;

    /// <summary>Creates the collector over an optional PII-safe diagnostics sink.</summary>
    public AiProviderSectionDiagnostics(Action<string>? sink = null) => _sink = sink;

    /// <summary>Number of times the surface has been opened.</summary>
    public long ViewsOpened => Interlocked.Read(ref _viewsOpened);

    /// <summary>Record that the surface was opened, emitting <c>view.opened slug=AIProviderSection</c>.</summary>
    public void RecordViewOpened()
    {
        Interlocked.Increment(ref _viewsOpened);
        _sink?.Invoke($"view.opened slug={AiProviderSectionRegistration.Slug}");
    }
}

/// <summary>Whether a validation banner reports success or failure (drives its tone token).</summary>
public enum AiProviderBannerKind
{
    /// <summary>The provider was reachable.</summary>
    Ok,

    /// <summary>The configuration was rejected or the probe faulted.</summary>
    Fail,
}

/// <summary>An inline validation banner: a tone + a finished message. Pure data.</summary>
public sealed record AiProviderBanner(AiProviderBannerKind Kind, string Message);

/// <summary>One option in a provider / Azure-surface drop-down: the wire <see cref="Value"/> and its display
/// <see cref="Label"/>. Mirrors the web <c>Select</c> options arrays. Pure data.</summary>
public sealed record AiProviderOption(string Value, string Label);

File: apps/android/app/src/main/java/com/teslasync/feature-views/AIProviderSection/AIProviderSectionModel.kt
// Pure, framework-free model + projection for the AIProviderSection feature view — the native analogue of
// everything the web component derives before returning JSX
// (web/src/features/settings/components/AIProviderSection.tsx). No Compose, no Android, no HTTP: every
// declaration here is unit-tested off-device in the :android:testReleaseUnitTest gate, keeping the
// composable a thin render layer.
//
// AIProviderSection is the Settings → AI provider-configuration form. It is a *controlled* component: the
// owning settings page owns the draft ([AIProviderDraft]) and the cloud/local mode, passing them down with a
// single `onChange` setter; this surface owns only the ephemeral validate-banner state and the pre-flight
// `POST /settings/ai/validate-config` call. So — exactly like the web source — there is no cache-then-network
// query feed here: the one asynchronous action is the imperative validate mutation
// (`useValidateAiProvider` → the shared [io.teslasync.shared.core.presentation.aisettings.AiSettingsStore]
// `validateAiProvider`, bound through the S8 seam in AIProviderSectionSource). Because there is no fetched
// feed, the prompt's generic stale/offline freshness surfaces do not apply (the web source has none either);
// the surfaces this form does render are: the always-present field content, the idle / validating /
// success / failure validate states, and the transport-error failure the View raises for a non-422 error.
//
// Every render decision the web component computes inline — the provider/flavor option sets, the
// `provider === 'azure' && flavor !== 'foundry'` model-label switch, the Azure sub-panel guard, the
// cents↔dollars cost-cap mapping, the validate-request body (cloud full set vs local mode+url, with the
// `api_key` blank→omit rule), the button-enablement predicates, and the success-banner variant selection
// (pinned → probed → plain) — is reproduced here as a pure function so the View only resolves i18n
// (P1/S10) + design tokens (P1/S9) and draws what the projection returns.
//
// `InvalidPackageDeclaration` is suppressed because this surface's mandated directory
// (com/teslasync/feature-views/AIProviderSection — the P3 prompt's allowed-files path) cannot form a valid
// Kotlin package (a hyphen and a PascalCase segment are illegal in a package identifier), so the package
// intentionally diverges from the path. `MatchingDeclarationName` is suppressed for the co-located
// supporting declarations.
@file:Suppress("MatchingDeclarationName", "InvalidPackageDeclaration")

package io.teslasync.android.featureviews.aiprovider

import io.teslasync.shared.core.diagnostics.Logger
import io.teslasync.shared.core.presentation.aisettings.ValidateAiProviderRequest
import io.teslasync.shared.core.presentation.aisettings.ValidateAiProviderResult
import kotlin.math.roundToInt

/**
 * Local in-memory shape for the provider form — the native mirror of the web `AIProviderDraft`. Holds the
 * server's `ai_provider_config` fields plus the sibling cost cap so the parent can pass a single setter.
 * Kotlin uses camelCase for the field names; the SI/wire mapping back to the snake_case request body is
 * done in [AIProviderProjection.buildValidateRequest].
 *
 * @property provider provider id (`openai`/`anthropic`/`azure`/`google` in cloud, `ollama`/`lmstudio`/
 *   `llama-cpp` in local).
 * @property baseUrl provider base URL (local Ollama-style URL, or the Azure resource endpoint).
 * @property model model identifier used for routing / cost tracking.
 * @property apiKey lives in memory only; an empty string means "no change on save" and is omitted from the
 *   validate body so the backend falls back to the saved (encrypted) key.
 * @property costCapCents daily cap in cents; `0` means unset.
 * @property apiVersion Azure-only API-version query parameter.
 * @property flavor Azure-only surface selector (`openai` = Azure OpenAI Service, `foundry` = Azure AI
 *   Foundry / Inference).
 * @property deployment Azure-only chat deployment name; blank → the adapter falls back to [model].
 * @property embeddingModel embedding model identifier (used by the RAG worker).
 * @property embeddingDeployment Azure-only embedding deployment name.
 */
data class AIProviderDraft(
    val provider: String = "",
    val baseUrl: String = "",
    val model: String = "",
    val apiKey: String = "",
    val costCapCents: Int = 0,
    val apiVersion: String = "",
    val flavor: String = "",
    val deployment: String = "",
    val embeddingModel: String = "",
    val embeddingDeployment: String = "",
)

/**
 * One selectable provider/flavor option — a stable [value] and a display [label]. The labels are vendor
 * brand names / product-surface descriptors that the web source renders verbatim (never through `t()`), so
 * they are carried as literals here for faithful parity rather than routed through the i18n catalog.
 */
data class ProviderChoice(
    val value: String,
    val label: String,
)

/** Whether a completed validation succeeded or was rejected — drives the banner's accent color. */
enum class ValidateOutcome { Ok, Fail }

/**
 * The render-ready validate-banner message, kept i18n-free so the projection stays framework-free: the View
 * resolves each case to a localized string (P1/S10). [Failure] carries the server-provided human message
 * verbatim (the web `result.message`), which is already localized server-side and not a client catalog key.
 */
sealed interface ValidateMessage {
    /** Local validator pinned a resolved private IP — web `successPinned` ("OK — pinned to {{ip}}"). */
    data class Pinned(
        val ip: String,
    ) : ValidateMessage

    /** Cloud probe exercised a concrete model — web `successProbed` ("OK — {{model}} reachable"). */
    data class Probed(
        val model: String,
    ) : ValidateMessage

    /** Plain success with no pinned IP or probed model — web `success` ("OK — provider reachable"). */
    data object Reachable : ValidateMessage

    /** A rejection or transport failure carrying the message to display (web `result.message`). */
    data class Failure(
        val text: String,
    ) : ValidateMessage
}

/** The fully projected validate banner shown after a validation completes; `null` means no banner. */
data class ValidateBanner(
    val outcome: ValidateOutcome,
    val message: ValidateMessage,
)

/**
 * Pure projection from the [AIProviderDraft] + cloud/local mode + validation result to everything the web
 * component computes inline. Stateless and framework-free; the single source of truth for the surface's
 * render decisions so they are verified once, off-device.
 */
object AIProviderProjection {
    /** Cloud provider choices (web cloud `options`). Labels are vendor brand names, rendered verbatim. */
    val cloudProviderChoices: List<ProviderChoice> =
        listOf(
            ProviderChoice("openai", "OpenAI"),
            ProviderChoice("anthropic", "Anthropic"),
            ProviderChoice("azure", "Azure AI"),
            ProviderChoice("google", "Google"),
        )

    /** Local provider choices (web local `options`). Labels are product brand names, rendered verbatim. */
    val localProviderChoices: List<ProviderChoice> =
        listOf(
            ProviderChoice("ollama", "Ollama"),
            ProviderChoice("lmstudio", "LM Studio"),
            ProviderChoice("llama-cpp", "llama.cpp"),
        )

    /** The Azure surface choices (web Azure-flavor `options`). Descriptors rendered verbatim like the web. */
    val azureFlavorChoices: List<ProviderChoice> =
        listOf(
            ProviderChoice("openai", "Azure OpenAI Service (gpt-4o, gpt-4-turbo, …)"),
            ProviderChoice("foundry", "Azure AI Foundry / Inference (multi-vendor)"),
        )

    /** The provider choices for the active [isCloud] mode (web `isCloud ? cloud : local`). */
    fun providerChoices(isCloud: Boolean): List<ProviderChoice> = if (isCloud) cloudProviderChoices else localProviderChoices

    /**
     * Whether the Azure sub-panel (surface + api-version + deployment fields) and the Azure resource-endpoint
     * URL are shown — web `isCloud && value.provider === 'azure'`.
     */
    fun showAzureConfig(
        isCloud: Boolean,
        provider: String,
    ): Boolean = isCloud && provider == AZURE

    /**
     * Whether the Azure deployment-name fields (chat + embedding) are shown — web `value.flavor !== 'foundry'`
     * within the Azure sub-panel (the Foundry surface routes the model in the body, so it has no deployment).
     */
    fun showAzureDeploymentFields(flavor: String): Boolean = flavor != FOUNDRY

    /**
     * Whether the Model field uses the Azure "model identifier" label + cost-tracking hint, rather than the
     * plain "Model" label — web `value.provider === 'azure' && value.flavor !== 'foundry'`.
     */
    fun usesAzureModelLabel(
        provider: String,
        flavor: String,
    ): Boolean = provider == AZURE && flavor != FOUNDRY

    /** The Azure-surface select value, defaulting blank → `openai` (web `value.flavor || 'openai'`). */
    fun azureFlavorValue(flavor: String): String = flavor.ifBlank { OPENAI }

    /** Model-field example (ghost prompt) — web `isCloud ? 'gpt-4o-mini' : 'llama3.1:8b'`. */
    fun modelExample(isCloud: Boolean): String = if (isCloud) "gpt-4o-mini" else "llama3.1:8b"

    /** Azure chat-deployment example (ghost prompt) — web `value.model || 'gpt-4o-mini'`. */
    fun azureDeploymentExample(model: String): String = model.ifBlank { "gpt-4o-mini" }

    /** Azure embedding-deployment example (ghost prompt) — web `value.embedding_model || 'text-embedding-3-small'`. */
    fun azureEmbeddingExample(embeddingModel: String): String = embeddingModel.ifBlank { "text-embedding-3-small" }

    /**
     * The cost-cap field's display string — web `cost_cap_cents > 0 ? (cents / 100).toFixed(2) : ''`. Uses
     * exact integer arithmetic (no float rounding, no locale grouping) so it round-trips with
     * [parseCostCapCents]: 500 → "5.00", 5 → "0.05", 100000 → "1000.00", 0 → "".
     */
    fun costCapDisplay(cents: Int): String {
        if (cents <= 0) return ""
        val whole = cents / CENTS_PER_UNIT
        val frac = cents % CENTS_PER_UNIT
        return "$whole.${frac.toString().padStart(2, '0')}"
    }

    /**
     * Parses a dollars string back to cents — web `const dollars = parseFloat(v); isFinite ? max(0,
     * round(dollars * 100)) : 0`. A non-numeric or empty value yields `0`; negatives clamp to `0`.
     */
    fun parseCostCapCents(input: String): Int {
        val dollars = input.trim().toDoubleOrNull() // parity:allow stdlib parser; "toDo" substring false positive
        if (dollars == null || !dollars.isFinite()) return 0
        return (dollars * CENTS_PER_UNIT).roundToInt().coerceAtLeast(0)
    }

    /**
     * Builds the `POST /settings/ai/validate-config` request body for the active mode — the verbatim port of
     * the web `validate.mutateAsync(...)` payload.
     *
     * Local mode sends only `mode`/`provider`/`base_url` (the validator is provider-agnostic there). Cloud
     * mode sends the full configuration so the backend can build a real adapter and run a 1-token probe;
     * `api_key` is omitted (null) when the user did not type one so the backend falls back to the saved
     * (encrypted) key rather than clobbering it with "", while the remaining fields are sent as-is (empty
     * strings included) exactly as the web `JSON.stringify` body does.
     */
    fun buildValidateRequest(
        draft: AIProviderDraft,
        isCloud: Boolean,
    ): ValidateAiProviderRequest =
        if (isCloud) {
            ValidateAiProviderRequest(
                mode = MODE_CLOUD,
                provider = draft.provider,
                baseUrl = draft.baseUrl,
                apiKey = draft.apiKey.takeIf { it.trim().isNotEmpty() },
                model = draft.model,
                apiVersion = draft.apiVersion,
                flavor = draft.flavor,
                deployment = draft.deployment,
                embeddingModel = draft.embeddingModel,
                embeddingDeployment = draft.embeddingDeployment,
            )
        } else {
            ValidateAiProviderRequest(
                mode = MODE_LOCAL,
                provider = draft.provider,
                baseUrl = draft.baseUrl,
            )
        }

    /**
     * Maps a completed [ValidateAiProviderResult] to the render-ready [ValidateBanner] — web's
     * `result.ok ? (pinned_ip ? successPinned : probed_model ? successProbed : success) : { fail, message }`.
     * Blank `pinnedIp`/`probedModel` are treated as absent to mirror JS truthiness (empty string is falsy).
     */
    fun bannerFrom(result: ValidateAiProviderResult): ValidateBanner =
        when (result) {
            is ValidateAiProviderResult.Success -> {
                val pinned = result.pinnedIp?.takeIf { it.isNotBlank() }
                val probed = result.probedModel?.takeIf { it.isNotBlank() }
                val message =
                    when {
                        pinned != null -> ValidateMessage.Pinned(pinned)
                        probed != null -> ValidateMessage.Probed(probed)
                        else -> ValidateMessage.Reachable
                    }
                ValidateBanner(ValidateOutcome.Ok, message)
            }

            is ValidateAiProviderResult.Failure -> ValidateBanner(ValidateOutcome.Fail, ValidateMessage.Failure(result.message))
        }

    /**
     * Whether the local Validate button is enabled — web `!validate.isPending && base_url.trim().length > 0`.
     * (The button's spinner/disabled state while pending is driven separately by the loading flag.)
     */
    fun localValidateEnabled(baseUrl: String): Boolean = baseUrl.trim().isNotEmpty()

    private const val AZURE = "azure"
    private const val FOUNDRY = "foundry"
    private const val OPENAI = "openai"
    private const val MODE_CLOUD = "cloud"
    private const val MODE_LOCAL = "local"
    private const val CENTS_PER_UNIT = 100
}

/**
 * The one PII-safe diagnostic this surface emits (P1/S11). Carries only the surface [SLUG] — never the
 * provider URL, API key, model, or any other field — so a diagnostics line can never leak provider
 * configuration or secrets.
 */
object AIProviderSectionDiagnostics {
    /** Diagnostics surface slug emitted with the `view.opened` event. */
    const val SLUG: String = "AIProviderSection"

    private const val VIEW_OPENED = "view.opened"
    private const val SURFACE_KEY = "surface"

    /** Emits the `view.opened` diagnostic for this surface. Call from the composable's first-composition effect. */
    fun recordViewOpened(logger: Logger) {
        logger.info(VIEW_OPENED, mapOf(SURFACE_KEY to SLUG))
    }
}
